### ISSUE TITLE
Fix slow tests

### ISSUE THREAD

**Issue by zcorpan** [web-platform-tests/wpt#11571]
From https://github.com/web-platform-tests/wpt/issues/9972#issuecomment-398058137

 - [ ] 23.2075 xhr/xmlhttprequest-timeout-worker-synconworker.html
 - [ ] 23.0962 xhr/xmlhttprequest-timeout-worker-overrides.html
 - [ ] 23.0917 xhr/xmlhttprequest-timeout-worker-simple.html
 - [ ] 23.0835 xhr/xmlhttprequest-timeout-worker-twice.html
 - [ ] 23.0772 xhr/xmlhttprequest-timeout-simple.html
 - [ ] 23.0693 xhr/xmlhttprequest-timeout-twice.html
 - [ ] 23.0654 xhr/xmlhttprequest-timeout-overrides.html
 - [ ] 22.0524 xhr/xmlhttprequest-timeout-worker-aborted.html
 - [ ] 22.0496 xhr/xmlhttprequest-timeout-aborted.html
 - [ ] 21.0636 xhr/xmlhttprequest-timeout-overridesexpires.html
 - [ ] 21.0554 xhr/xmlhttprequest-timeout-worker-overridesexpires.html
 - [ ] 20.2750 websockets/keeping-connection-open/001.html
 - [ ] 20.0833 websockets/keeping-connection-open/001.html?wss
 - [ ] 14.5754 html/semantics/embedded-content/the-img-element/sizes/parse-a-sizes-attribute.html
 - [ ] 14.1288 xhr/sync-no-timeout.any.html
 - [ ] 14.1202 xhr/sync-no-timeout.any.sharedworker.html
 - [ ] 14.1000 xhr/sync-no-timeout.any.worker.html
 - [ ] 12.6278 fetch/api/request/request-keepalive-quota.html
 - [ ] 12.3427 webrtc/RTCDTMFSender-ontonechange-long.https.html
 - [ ] 11.1061 fetch/http-cache/cc-request.html
 - [ ] 11.0543 xhr/xmlhttprequest-timeout-abortedonmain.html
 - [ ] 10.7306 html/semantics/embedded-content/media-elements/video_loop_base.html
 - [ ] 10.3728 html/semantics/embedded-content/media-elements/audio_loop_base.html
 - [ ] 10.1424 service-workers/service-worker/fetch-frame-resource.https.html
 - [ ] 10.0538 xhr/xmlhttprequest-timeout-synconmain.html
 - [ ] 08.2459 html/semantics/scripting-1/the-script-element/async_007.htm
 - [ ] 08.2032 websockets/constructor/014.html
 - [ ] 08.1276 html/semantics/scripting-1/the-script-element/async_010.htm
 - [ ] 08.1174 websockets/constructor/014.html?wss
 - [ ] 07.8624 offscreen-canvas/convert-to-blob/offscreencanvas.convert.to.blob.html
 - [ ] 07.4726 html/semantics/document-metadata/the-meta-element/pragma-directives/attr-meta-http-equiv-refresh/parsing.html
 - [ ] 07.4439 html/semantics/forms/textfieldselection/select-event.html
 - [ ] 07.1871 fetch/http-cache/freshness.html
 - [ ] 06.5026 beacon/beacon-basic-formdataMax.html
 - [ ] 06.2166 editing/run/justifycenter.html
 - [ ] 06.0157 editing/run/multitest.html
 - [ ] 06.0153 IndexedDB/interleaved-cursors-large.html
 - [ ] 05.8177 beacon/beacon-basic-formdata.html
 - [ ] 05.7452 fetch/api/redirect/redirect-referrer.any.html
 - [ ] 05.6613 html/semantics/embedded-content/media-elements/track/track-element/track-cues-cuechange.html
 - [ ] 05.6235 html/semantics/embedded-content/media-elements/track/track-element/track-cues-enter-exit.html
 - [ ] 05.5950 pointerevents/pointerevent_pointerout_received_once-manual.html
 - [ ] 05.5754 html/semantics/scripting-1/the-script-element/execution-timing/078.html
 - [ ] 05.4589 content-security-policy/generic/no-default-src.sub.html
 - [ ] 05.3875 content-security-policy/style-src/style-src-multiple-policies-multiple-hashing-algorithms.html
 - [ ] 05.3707 content-security-policy/reporting/report-cross-origin-no-cookies.sub.html
 - [ ] 05.3616 preload/onerror-event.html
 - [ ] 05.3135 fetch/api/abort/general.any.worker.html
 - [ ] 05.3042 pointerevents/pointerevent_touch-action-svg-test_touch-manual.html
 - [ ] 05.2828 WebCryptoAPI/derive_bits_keys/test_pbkdf2_long_empty.https.html
 - [ ] 05.2788 preload/link-header-preload.html
 - [ ] 05.2624 preload/preload-csp.sub.html
 - [ ] 05.2121 fetch/api/redirect/redirect-origin.any.html
 - [ ] 05.2014 cors/304.htm
 - [ ] 05.1825 html/semantics/scripting-1/the-script-element/async_005.htm
 - [ ] 05.1741 fetch/security/redirect-to-url-with-credentials.https.html
 - [ ] 05.1698 websockets/constructor/013.html?wss
 - [ ] 05.1680 preload/dynamic-adding-preload.html
 - [ ] 05.1651 preload/link-header-preload-nonce.html
 - [ ] 05.1536 preload/preload-default-csp.sub.html
 - [ ] 05.1536 preload/link-header-on-subresource.html
 - [ ] 05.1078 html/semantics/scripting-1/the-script-element/script-not-executed-after-shutdown.html
 - [ ] 05.0634 preload/delaying-onload-link-preload-after-discovery.html
 - [ ] 05.0386 preload/onload-event.html
 - [ ] 05.0338 preload/preload-with-type.html
 - [ ] 05.0105 pointerevents/pointerevent_touch-action-table-test_touch-manual.html
 - [ ] 05.0066 fetch/api/abort/general.any.sharedworker.html
 - [ ] 04.9596 fetch/api/cors/cors-redirect-credentials.any.worker.html
 - [ ] 04.8815 fetch/api/cors/cors-preflight-status.any.html
 - [ ] 04.7560 fetch/api/cors/cors-preflight-status.any.worker.html
 - [ ] 04.6807 fetch/api/abort/general.any.html
 - [ ] 04.6238 cookies/http-state/general-tests.html
 - [ ] 04.5852 service-workers/service-worker/fetch-response-taint.https.html
 - [ ] 04.5500 xhr/xmlhttprequest-timeout-reused.html
 - [ ] 04.5359 fetch/http-cache/invalidate.html
 - [ ] 04.5019 WebCryptoAPI/derive_bits_keys/test_pbkdf2_long_short.https.html
 - [ ] 04.4986 WebCryptoAPI/derive_bits_keys/test_pbkdf2_short_long.https.html
 - [ ] 04.4294 fetch/api/cors/cors-redirect.any.html
 - [ ] 04.4151 fetch/api/redirect/redirect-referrer.any.worker.html
 - [ ] 04.3665 WebCryptoAPI/derive_bits_keys/test_pbkdf2_short_empty.https.html
 - [ ] 04.3415 WebCryptoAPI/derive_bits_keys/test_pbkdf2_long_long.https.html
 - [ ] 04.3259 editing/run/justifyright.html
 - [ ] 04.3206 fetch/api/response/response-clone.html
 - [ ] 04.2876 fetch/http-cache/heuristic.html
 - [ ] 04.2819 dom/nodes/Document-characterSet-normalization.html
 - [ ] 04.2148 fetch/api/request/request-cache-force-cache.html
 - [ ] 04.2130 encoding/legacy-mb-tchinese/big5/big5-encode-form-errors-extBb.html?21001-last
 - [ ] 04.1920 editing/run/insertparagraph.html
 - [ ] 04.1794 html/dom/documents/resource-metadata-management/document-lastModified-01.html
 - [ ] 04.1786 html/semantics/scripting-1/the-script-element/async_004.htm
 - [ ] 04.1738 html/semantics/scripting-1/the-script-element/async_006.htm
 - [ ] 04.1716 fetch/api/request/request-cache-reload.html
 - [ ] 04.1643 html/semantics/embedded-content/the-iframe-element/sandbox_005.htm
 - [ ] 04.1583 html/semantics/scripting-1/the-script-element/async_003.htm
 - [ ] 04.1379 html/semantics/scripting-1/the-script-element/execution-timing/050.html
 - [ ] 04.1288 fetch/api/cors/cors-redirect-credentials.any.html
 - [ ] 04.1003 fetch/api/redirect/redirect-origin.any.worker.html
 - [ ] 04.0648 xhr/response-data-progress.htm
 - [ ] 03.9741 service-workers/service-worker/registration-updateviacache.https.html
 - [ ] 03.9057 encoding/legacy-mb-japanese/euc-jp/eucjp-encode-form-errors-han.html?21001-last
 - [ ] 03.9044 fetch/http-cache/vary.html
 - [ ] 03.8955 WebCryptoAPI/derive_bits_keys/test_pbkdf2_empty_long.https.html
 - [ ] 03.8817 html/browsers/origin/cross-origin-objects/cross-origin-objects.html
 - [ ] 03.8354 cookies/http-state/attribute-tests.html
 - [ ] 03.7963 cors/allow-headers.htm
 - [ ] 03.7916 editing/run/justifyfull.html
 - [ ] 03.7898 dom/ranges/Range-set.html
 - [ ] 03.7599 WebCryptoAPI/derive_bits_keys/test_pbkdf2_short_short.https.html
 - [ ] 03.7151 encoding/legacy-mb-japanese/iso-2022-jp/iso2022jp-encode-form-errors-han.html?19001-20000
 - [ ] 03.7150 dom/ranges/Range-compareBoundaryPoints.html
 - [ ] 03.7069 fetch/api/headers/header-values-normalize.html
 - [ ] 03.7064 websockets/constructor/013.html
 - [ ] 03.6490 WebCryptoAPI/derive_bits_keys/test_pbkdf2_empty_short.https.html
 - [ ] 03.5475 fetch/api/request/request-cache-only-if-cached.html
 - [ ] 03.5284 WebCryptoAPI/derive_bits_keys/test_pbkdf2_empty_empty.https.html
 - [ ] 03.4752 preload/single-download-preload.html
 - [ ] 03.4160 fetch/api/cors/cors-preflight-referrer.any.worker.html
 - [ ] 03.4107 fetch/api/redirect/redirect-location.any.html
 - [ ] 03.3984 encoding/single-byte-decoder.html?XMLHttpRequest
 - [ ] 03.3828 pointerevents/pointerevent_touch-action-span-test_touch-manual.html
 - [ ] 03.3617 eventsource/format-field-retry-bogus.htm
 - [ ] 03.3292 html/semantics/scripting-1/the-script-element/execution-timing/082.html
 - [ ] 03.3243 editing/run/formatblock.html
 - [ ] 03.3040 encoding/legacy-mb-tchinese/big5/big5-encode-form-errors-extBb.html?16001-17000
 - [ ] 03.2778 eventsource/format-field-retry.htm
 - [ ] 03.2644 service-workers/service-worker/navigation-redirect.https.html
 - [ ] 03.2574 cors/preflight-cache.htm
 - [ ] 03.2400 html/semantics/document-metadata/the-meta-element/pragma-directives/attr-meta-http-equiv-refresh/not-in-shadow-tree.html
 - [ ] 03.2301 fetch/api/redirect/redirect-method.any.worker.html
 - [ ] 03.2147 preload/link-header-preload-srcset.tentative.html
 - [ ] 03.2126 xhr/access-control-basic-allow-preflight-cache-timeout.htm
 - [ ] 03.2077 fetch/api/redirect/redirect-location.any.worker.html
 - [ ] 03.1888 html/semantics/document-metadata/the-link-element/link-load-event.html
 - [ ] 03.1748 fetch/api/cors/cors-preflight.any.worker.html
 - [ ] 03.1724 html/semantics/scripting-1/the-script-element/execution-timing/111.html
 - [ ] 03.1696 css/css-animations/animationevent-types.html
 - [ ] 03.1619 service-workers/service-worker/fetch-event-respond-with-body-loaded-in-chunk.https.html
 - [ ] 03.1534 html/semantics/scripting-1/the-script-element/execution-timing/112.html
 - [ ] 03.1481 html/semantics/document-metadata/interactions-of-styling-and-scripting/conditionally-block-rendering-on-link-media-attr.html
 - [ ] 03.1453 encoding/legacy-mb-japanese/iso-2022-jp/iso2022jp-encode-form-errors-han.html?21001-last
 - [ ] 03.1116 xhr/sync-no-progress.any.worker.html
 - [ ] 03.0933 payment-request/allowpaymentrequest/setting-allowpaymentrequest-timing.https.sub.html
 - [ ] 03.0857 xhr/sync-no-progress.any.html
 - [ ] 03.0575 xhr/event-progress.htm
 - [ ] 03.0574 xhr/progress-events-response-data-gzip.htm
 - [ ] 03.0509 html/semantics/scripting-1/the-script-element/execution-timing/015a.html
 - [ ] 03.0434 html/browsers/offline/application-cache-api/api_status_idle.https.html
 - [ ] 03.0127 encoding/legacy-mb-japanese/iso-2022-jp/iso2022jp-encode-form-errors-han.html?18001-19000
 - [ ] 03.0100 encoding/legacy-mb-japanese/euc-jp/eucjp-encode-form-errors-han.html?13001-14000
 - [ ] 02.9842 encoding/single-byte-decoder.html?document
 - [ ] 02.9755 encoding/legacy-mb-japanese/euc-jp/eucjp-encode-form-errors-han.html?18001-19000
 - [ ] 02.9745 content-security-policy/navigation/javascript-url-navigation-inherits-csp.html
 - [ ] 02.9614 media-source/mediasource-config-change-webm-v-bitrate.html
 - [ ] 02.9492 cors/simple-requests.htm
 - [ ] 02.9161 service-workers/service-worker/registration-mime-types.https.html
 - [ ] 02.9152 html/semantics/tabular-data/processing-model-1/span-limits.html
 - [ ] 02.9125 encoding/legacy-mb-korean/euc-kr/euckr-encode-form-windows-949.html?12001-13000
 - [ ] 02.9095 service-workers/service-worker/fetch-event-redirect.https.html
 - [ ] 02.9069 orientation-sensor/RelativeOrientationSensor.https.html
 - [ ] 02.9040 encoding/legacy-mb-japanese/euc-jp/eucjp-encode-form-errors-han.html?19001-20000
 - [ ] 02.9005 orientation-sensor/AbsoluteOrientationSensor.https.html
 - [ ] 02.8946 encoding/legacy-mb-japanese/euc-jp/eucjp-encode-form-errors-han.html?20001-21000
 - [ ] 02.8292 fetch/api/basic/request-headers.any.html
 - [ ] 02.8070 fetch/api/basic/request-forbidden-headers.any.worker.html
 - [ ] 02.7952 css/css-values/vh_not_refreshing_on_chrome.html
 - [ ] 02.7931 webrtc/RTCDTMFSender-ontonechange.https.html
 - [ ] 02.7913 encoding/legacy-mb-japanese/iso-2022-jp/iso2022jp-encode-form-errors-han.html?20001-21000
 - [ ] 02.7814 dom/ranges/Range-mutations-dataChange.html
 - [ ] 02.7679 fetch/api/redirect/redirect-method.any.html
 - [ ] 02.7654 media-source/mediasource-config-change-webm-av-audio-bitrate.html
 - [ ] 02.7426 fetch/api/abort/general.https.any.serviceworker.html
 - [ ] 02.7221 fetch/api/cors/cors-no-preflight.any.html
 - [ ] 02.7191 fetch/api/basic/request-headers.any.worker.html
 - [ ] 02.7112 encoding/legacy-mb-japanese/shift_jis/sjis-encode-form-errors-han.html?17001-18000
 - [ ] 02.6968 encoding/legacy-mb-korean/euc-kr/euckr-encode-form-ksc_5601.html?15001-16000
 - [ ] 02.6879 media-source/mediasource-config-change-webm-av-framesize.html
 - [ ] 02.6841 media-source/mediasource-config-change-webm-v-framerate.html
 - [ ] 02.6772 encoding/legacy-mb-korean/euc-kr/euckr-encode-form-ksc_5601.html?12001-13000
 - [ ] 02.6744 fetch/api/basic/request-forbidden-headers.any.html

**Comment by zcorpan** [comment]
> 20.2750 websockets/keeping-connection-open/001.html
> 20.0833 websockets/keeping-connection-open/001.html?wss

This takes 20s by design; the only fix would be to remove the test.

It doesn't show up in https://bocoup.github.io/wpt-disabled-tests-report/#filter=websockets%2Fkeeping-connection-open%2F001.html so I'm assuming test results are stable.

**Comment by zcorpan** [comment]
> 12.3427 webrtc/RTCDTMFSender-ontonechange-long.https.html

I think this needs to take this long for what it's testing (a bit unclear why it takes 12 seconds when the timeout is 6 seconds, though).

**Comment by foolip** [comment]
I'll be OOO until July 5, but I'll check back in here (some time) after I'm back.

**Comment by wanderview** [comment]
> 10.1424 service-workers/service-worker/fetch-frame-resource.https.html

I don't think splitting would help here.  It only has a handful of cases and they all run in parallel.

The reason it can be slow is that it uses a 10 second setTimeout() to detect failed cross-origin window loads.  AFAIK there is no cross-browser way to handle this case and the timeout was the best attempt to deal with it.  If we want this test to be faster then we need another solution for detecting this case that works in all browsers.

**Comment by wanderview** [comment]
Also, it feels wrong to have some kind of arbitrary point where we require splitting tests.  What is really driving this?  It may be slower in some browsers than others, but these are not performance tests, etc.

Also, there seems to be a trend towards more new tests being written as promise_test() which serializes execution.  In contrast, we could encourage folks to write async_test() instead that runs test cases in parallel.  Or we should make a promise test method that runs in parallel.

Anyway, this splitting thing feels like its pointing towards some kind of infrastructure deficiency or something.  It doesn't seem like splitting should be necessary or will be scalable in the long term.

**Comment by zcorpan** [comment]
If the test needs to take a long time, tick the checkbox and document rationale in a comment (as you did).

In #9972 @ricea said:

> Chromium developers are spending a lot of time triaging test flakiness caused by tests that are simply too slow.

I'm trying to address this by investigating the slowest tests (this issue) and [tests that are timing out or are flaky](https://bocoup.github.io/wpt-disabled-tests-report/), as well as implementing a [check to reject tests that are near their timeout](https://github.com/web-platform-tests/wpt/pull/11570). Sometimes splitting tests can lead to more stable results, sometimes not.

**Comment by zcorpan** [comment]
> 04.2130 encoding/legacy-mb-tchinese/big5/big5-encode-form-errors-extBb.html?21001-last

encoding tests have already been split up.

**Comment by wanderview** [comment]
Sure, but I also saw tests being split in chromium issues where the `<meta name="timeout" content="long">` could have been added.  I also feel like they've added a lot of serialized promise tests instead of parallel async tests.  Both of those are strategies that could be pursued before test splitting.

Also, I would note test splitting creates a lot of churn in history, test expectations, and greater chance of conflict during automatic sync'ing.

**Comment by domenic** [comment]
> @domenic can you split this up using splitTestByKey similarly to #11593 (1 key per element)?

I don't have the time to devote to this at the moment, but will be happy to review any such changes someone else is able to make.
